sdk/src/simple_state/addresser.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crypto::digest::Digest;
+use crypto::sha2::Sha512;
+
+pub trait Addresser<K> {
+    /// Returns a radix address calculated from the given keys
+    ///
+    /// # Arguments
+    ///
+    /// * `keys` - Contains natural keys used to calculate an address
+    ///
+    fn compute(&self, keys: &K) -> String;
+
+    /// Returns a human readable string of the given keys
+    ///
+    /// # Arguments
+    ///
+    /// * `keys` - Contains natural keys
+    ///
+    fn normalize(&self, keys: &K) -> String;
+}
+
+fn hash(key: &str) -> String {
+    let mut sha = Sha512::new();
+    sha.input(key.as_bytes());
+    sha.result_str().to_string()
+}
+
+pub struct KeyHashAddresser {
+    prefix: String,
+}
+
+impl KeyHashAddresser {
+    pub fn new(prefix: String) -> KeyHashAddresser {
+        KeyHashAddresser { prefix }
+    }
+}
+
+impl Addresser<String> for KeyHashAddresser {
+    fn compute(&self, keys: &String) -> String {
+        let mut sha = Sha512::new();
+        sha.input(keys.as_bytes());
+        let hash_length = 70 - self.prefix.len();
+
+        String::from(&self.prefix) + &sha.result_str()[..hash_length].to_string()
+    }
+
+    fn normalize(&self, key: &String) -> String {
+        key.to_string()
+    }
+}
+
+pub struct DoubleKeyHashAddresser {
+    prefix: String,
+}
+
+impl DoubleKeyHashAddresser {
+    pub fn new(prefix: String) -> DoubleKeyHashAddresser {
+        DoubleKeyHashAddresser { prefix }
+    }
+}
+
+impl Addresser<(String, String)> for DoubleKeyHashAddresser {
+    fn compute(&self, keys: &(String, String)) -> String {
+        let hash_length = 70 - self.prefix.len();
+        let first_hash_length = hash_length / 2;
```

Comment:
We need to be able to set all the hash lengths except the last one (which can be calculated). This could be in the constructor (or maybe an alternate constructor, with the simple one providing a default) or a set method.

For example, if we have (key1, key2), the number of key1s might be 10 and the number of key2s might be 10,000. In that case, we probably want the length of the first hash to be much shorter than the hash length used for the second, so we get better distribution in the state tree.

Reply:
I added an optional argument to the constructor for each of these structs